Refs #34976 -- Print directory structure after startapp and startproject

TobeTek opened the pull request:
Print the newly created directory structure after running the startapp and startproject commands.
This provides feedback that a command has successfully run.

Refs #34976

*Sample output*:
![Screenshot from 2024-02-15 12-51-01](https://github.com/django/django/assets/82294671/2455dffc-e2df-4f63-8136-80888e2546af)
![Screenshot from 2024-02-15 12-51-44](https://github.com/django/django/assets/82294671/fbb48964-906d-4656-b9dc-808496aa5829)

TobeTek committed:
Refs #34976 -- Print directory structure after startapp and startproject commands

Print the newly created directory structure after running the startapp and startproject commands.

Refs #34976

Use `make_style`  to style directories

Support --no-color flag in get_directory_tree

TobeTek committed:
Add Emmanuel T. Katchy to AUTHORS

felixxm commented:
Thanks for this patch, however, personally, I don't think it's worth adding. I accepted ticket-34976 for the simplest congrats message (#17722).

thibaudcolas commented:
Hi @felixxm, there’s rough agreement on the value of the directory structure in the [forum thread](https://forum.djangoproject.com/t/feature-request-startproject-and-startapp-should-provide-feedback/24073), and I added it to [ticket #34976](https://code.djangoproject.com/ticket/34976) as well. If you don’t think it’s worth adding I’d suggest making the case for that on the forum thread, as that’s where the output was discussed most, and the basis for me suggesting both the success message and the file structure output on the ticket.

felixxm commented:
> Hi @felixxm, there’s rough agreement on the value of the directory structure in the [forum thread](https://forum.djangoproject.com/t/feature-request-startproject-and-startapp-should-provide-feedback/24073), and I added it to [ticket #34976](https://code.djangoproject.com/ticket/34976) as well. If you don’t think it’s worth adding I’d suggest making the case for that on the forum thread, as that’s where the output was discussed most, and the basis for me suggesting both the success message and the file structure output on the ticket.

Whatever you prefer. This is just my opinion and is not more important than anyone else's.

thibaudcolas commented on django/core/management/utils.py:
```diff
@@ -173,3 +174,56 @@ def run_formatters(written_files, black_path=(sentinel := object())):
+    ...     print(line)
+    """
+
+    path = Path(path)
```
Is this needed? I’m surprised by this as it seems we’d be calling `get_directory_tree` with `Path` instances anyway, and in your example above we’re also calling it with a `Path`.

thibaudcolas commented on django/core/management/utils.py:
```diff
@@ -173,3 +174,56 @@ def run_formatters(written_files, black_path=(sentinel := object())):
+
+    path = Path(path)
+    if depth == 0:
+        yield colorize(path.name, fg="blue", opts=["bold"])
```
Cool stuff! I’d be interested to hear why you settled on those styles. 

And this might need further work to support Django’s `--no-color` CLI flag, and `DJANGO_COLORS` environment variable, and checks for color support? See `django.core.management.color.color_style`.

TobeTek replied:
I actually tried to mimic the colors/style I got from `tree` as closely as possible. In hindsight, maybe my zsh config might have had a hand to play :P

I've made the changes to support user color config, the only hitch is supporting the `--no-color` flag. I really think `django.core.management.color.Style` should have a property/attribute `is_no_color`? It would make things a tad bit easier.

thibaudcolas reviewed (commented):
Well now this is neat!

![django folder output](https://github.com/django/django/assets/877585/d4a26269-97bb-4301-9a9e-0b56728f0480)

TobeTek commented on django/core/management/utils.py:
```diff
@@ -173,3 +174,56 @@ def run_formatters(written_files, black_path=(sentinel := object())):
+def get_directory_tree(path, max_depth=10, prefix="", depth=0):
+    """
+    Get the folder structure of a directory and it's content,
+    akin to the GNU `tree` command.
```
No it isn't ([not on the list](https://en.wikipedia.org/wiki/List_of_GNU_Core_Utilities_commands))

nessita commented:
Hello @TobeTek! Would you have time to keep working on this? I see that the latest failure in windows is related to changes in this PR:
```
======================================================================
FAIL: test_startapp_unicode_name (admin_scripts.tests.DjangoAdminSettingsDirectory.test_startapp_unicode_name)
startapp creates the correct directory with Unicode characters.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\unittest\case.py", line 58, in testPartExecutor
    yield
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\unittest\case.py", line 634, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\unittest\case.py", line 589, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "D:\a\django\django\tests\admin_scripts\tests.py", line 758, in test_startapp_unicode_name
    self.assertNoOutput(err)
    ^^^^^^^^^^^^^^^^^
  File "D:\a\django\django\tests\admin_scripts\tests.py", line 171, in assertNoOutput
    self.assertEqual(
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\unittest\case.py", line 885, in assertEqual
    assertion_func(first, second, msg=msg)
    ^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\unittest\case.py", line 878, in _baseAssertEqual
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^^^
AssertionError: 1536 != 0 : Stream should be empty: actually contains 'Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "D:\a\django\django\django\__main__.py", line 10, in <module>
    management.execute_from_command_line()
  File "D:\a\django\django\django\core\management\__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "D:\a\django\django\django\core\management\__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "D:\a\django\django\django\core\management\base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "D:\a\django\django\django\core\management\base.py", line 461, in execute
    output = self.handle(*args, **options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\a\django\django\django\core\management\commands\startapp.py", line 14, in handle
    super().handle("app", app_name, target, **options)
  File "D:\a\django\django\django\core\management\templates.py", line 235, in handle
    self.stdout.write(subpath)
  File "D:\a\django\django\django\core\management\base.py", line 181, in write
    self._out.write(style_func(msg))
  File "C:\hostedtoolcache\windows\Python\3.12.2\x64\Lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeEncodeError: 'charmap' codec can't encode characters in position 0-4: character maps to <undefined>
'
```

/remind me to check on progress in one month

TobeTek commented:
Hi @nessita 
Happy to keep working on this.

I did [a bit of wrangling earlier](https://github.com/django/django/pull/17722#discussion_r1491797572), and I think [setting some additional environment variables](https://stackoverflow.com/a/63573649) might be a solution.

nessita commented:
> Hi @nessita Happy to keep working on this.
> 
> I did [a bit of wrangling earlier](https://github.com/django/django/pull/17722#discussion_r1491797572), and I think [setting some additional environment variables](https://stackoverflow.com/a/63573649) to deal with Unicode chars on Windows might be a solution.

Thanks for these details. I have pinged @smithdc1 about this in the context of PR #17722, because Salvo does not have a Windows system to use for debugging (neither do I). Do you think it could be beneficial if you and David sync efforts in untangling the Windows part of this challenge?

TobeTek commented:
> Do you think it could be beneficial if you and David sync efforts in untangling the Windows part of this challenge?

Happy to help out!

smithdc1 committed:
Early adopt UTF-8 Mode on Windows.

Python 3.15 will enable UTF-8 mode by default

https://peps.python.org/pep-0686/#enable-utf-8-mode-by-default

sarahboyce commented on django/core/management/utils.py:
```diff
@@ -173,3 +174,64 @@ def run_formatters(written_files, black_path=(sentinel := object())):
         )
+
+
+SPACE_PREFIX = "    "
+BRANCH_PREFIX = "|   "
+TEE_PREFIX = "|___ "
+ELBOW_PREFIX = "|___ "
```
I think I would put these inside the function
I also see no difference (or very very little difference) between TEE and ELBOW

sarahboyce commented on django/core/management/utils.py:
```diff
@@ -173,3 +174,64 @@ def run_formatters(written_files, black_path=(sentinel := object())):
+
+        is_last_element = index == len(dir_contents) - 1
+
+        # The last element in the tree gets a different prefix
+        if is_last_element:
+            yield prefix + ELBOW_PREFIX + subpath_name
+        else:
+            yield prefix + TEE_PREFIX + subpath_name
```
```suggestion
         yield prefix + ELBOW_PREFIX + subpath_name
```
I feel like TEE and ELBOW are the same?

sarahboyce commented on django/core/management/utils.py:
```diff
@@ -173,3 +174,64 @@ def run_formatters(written_files, black_path=(sentinel := object())):
+        options.get("force_color", False)
+    ):
+        directory_style = make_style("notice=blue,bold;")
+    else:
+        directory_style = no_style()
```
This is untested

sarahboyce commented on tests/admin_scripts/tests.py:
```diff
@@ -2405,6 +2406,25 @@ def test_wrong_args(self):
+
+                # We only check for a subset of the directory structure
+                self.assertOutput(out, name_project)
+                self.assertIn(f"{ELBOW_PREFIX}manage.py", out)
```
```suggestion
                self.assertIn("|___ manage.py", out)
```
I wouldn't use the constants in tests

sarahboyce commented on tests/admin_scripts/tests.py:
```diff
@@ -2405,6 +2406,25 @@ def test_wrong_args(self):
+                    args.append(dir)
+                out, err = self.run_django_admin(args)
+
+                # We only check for a subset of the directory structure
```
I think it's better to assert all of it? 🤔

TobeTek replied:
I was thinking, since the lower level tests verified most of the function's behaviour, another complete check would be too much?
Then again it does make more sense seeing the full directory structure when testing.

sarahboyce commented on tests/user_commands/tests.py:
```diff
@@ -535,3 +542,74 @@ def test_is_ignored_path_false(self):
+            ]
+            expected = [
+                format_dir(Path(tmpdir).name),
+                f"{TEE_PREFIX}file1.txt",
+                f"{TEE_PREFIX}file2.txt",
+                f"{ELBOW_PREFIX}{format_dir('foo')}",
+                f"{SPACE_PREFIX}{TEE_PREFIX}{format_dir('bar')}",
+                f"{SPACE_PREFIX}{BRANCH_PREFIX}{TEE_PREFIX}bar1.txt",
+                f"{SPACE_PREFIX}{BRANCH_PREFIX}{ELBOW_PREFIX}bar2.txt",
+                f"{SPACE_PREFIX}{TEE_PREFIX}foo1.txt",
+                f"{SPACE_PREFIX}{ELBOW_PREFIX}foo2.txt",
```
Please remove the constants, I think that would make this easier to read and see

sarahboyce reviewed (commented):
@TobeTek thank you for this 👍 

The comments I have in https://github.com/django/django/pull/17722 are relevant here about the missing release notes and docs
Not sure how you both want to manage this? It would definitely be easier to have 1 PR with co-authors 🤔

TobeTek commented:
Thanks for the feedback, @sarahboyce.
I'll be wrapping up exams this week and hope to apply the fixes over the weekend.

sarahboyce commented:
@TobeTek before you invest too much more work in this, want to let you know about the discussion on this feature: https://forum.djangoproject.com/t/feature-request-startproject-and-startapp-should-provide-feedback/24073/11
It's been raised that tree structures are difficult to read for screenreaders and I'd like the @django/accessibility team to confirm on what should be done here

TobeTek commented:
Thanks for the heads up! I'm going to go ahead and implement the changes you pointed out, and see where it goes from there 👍🏾
